### PR TITLE
Blog post TPI Jupyter

### PR DESCRIPTION
**Pull request by RCdeWit**
Some more padding to @casperdcl's blogpost. Kept the example/repo the same, which means it does provision a Tensorboard instance that is not really the focus of this blog post.

Goal: easy to follow tutorial for data scientists who're googling "jupyter notebooks in cloud".

- [x]  Is the content technically correct (examples are working and can be reproduced, code in the repo is up to date, etc.)
- [x]  There are no grammatical errors, jargon, or obvious stylistic issues.
- [x]  The flow and messaging correspond to what has been defined in the Phase 3.
- [ ]  Have images been run through tiny.png?
- [x]  Are all lines 80 characters or less?
- [x]  Are Oxford commas observed?
- [ ]  Content Creator has done a complete review for typos/language after taking a break from content
- [x]  links checked in piece
- [x]  Have reviewed what the piece looks like on your own [machine using yarn develop](https://dvc.org/doc/user-guide/contributing/docs#development-environment)

### CODE REVIEW

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+The second part of the configuration is perhaps the most interesting. Here we
+provide the specifications for the instance Terraform should provision for us.
+In this case, we would like a medium CPU and an NVIDIA T4 GPU, along with a
+125GB disk. Because machine types vary between cloud vendors,
+[TPI does some translation](https://registry.terraform.io/providers/iterative/iterative/latest/docs/resources/task#machine-type)
+from generic types (e.g., `s`/`m`/`l`/`xl`) to specific cloud machine types.
+This allows us to generalize these configurations and quickly switch from AWS to
+Azure, for example.
+
+Of particular interest here is the `spot = 0`, which tells TPI to provision spot
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:194]: given the blog's tone & level of detail in explanations, it implies a beginner target audience. Should thus explain briefly what spot/preemptible means. Pros/cons, give references.

It's also the No. 1 USP of TPI (auto-recovery from spot instance interrupts).

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+`spot = 0.12`), and TPI would hold off on provisioning until instances are
+available at that price. Disabling spot pricing (`spot = -1`) would let TPI
+provision instances at on-demand prices.
+
+</admon>
+
+After specifying the details for the hardware, we configure a few environment
+variables for the instance. We also set the local working directory for the
+script and the directory to to download files from upon terminating the instance
+(relative to the working directory). In this case, both are the `shared`
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:212]:
```suggestion
script and the directory to download files from (relative to the working
directory) upon terminating the instance. In this case, both are the `shared`
```

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+    while test ! -f log.md; do sleep 1; done
+    cat log.md
+    popd # dependencies
+
+    # start tensorboard in background
+    env -u JUPYTER_TOKEN -u AWS_ACCESS_KEY_ID -u AWS_SECRET_ACCESS_KEY -u REPO_TOKEN tensorboard --logdir . --host 0.0.0.0 --port 6006 &
+
+    # start Jupyter server in foreground
+    env -u AWS_ACCESS_KEY_ID -u AWS_SECRET_ACCESS_KEY -u REPO_TOKEN jupyter lab --allow-root --ip=0.0.0.0 --no-browser --port=8888 --port-retries=0
+  END
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:259]: borked indentation. Subtract 2 spaces

```suggestion
#!/bin/bash
  set -euo pipefail
  if test "$GITHUB_USER" != username; then
    # SSH debugging
    trap 'echo script error: waiting for debugging over SSH. Run \"terraform destroy\" to stop waiting; sleep inf' ERR
    mkdir -p "$HOME/.ssh"
    curl -fsSL "https://github.com/$GITHUB_USER.keys" >> "$HOME/.ssh/authorized_keys"
  fi
  export CUDACXX=/usr/local/cuda/bin/nvcc
  export DEBIAN_FRONTEND=noninteractive
  sed -ri 's#^(APT::Periodic::Unattended-Upgrade).*#\1 "0";#' /etc/apt/apt.conf.d/20auto-upgrades
  dpkg-reconfigure unattended-upgrades
  # install dependencies
  pip3 install $${QUIET:+-q} jupyterlab notebook matplotlib ipywidgets tensorflow==2.8.0 tensorboard tensorflow_datasets
  (curl -fsSL https://deb.nodesource.com/setup_16.x | bash -) >/dev/null
  apt-get install -y $${QUIET:+-qq} nodejs

  # start tunnel
  export JUPYTER_TOKEN="$(uuidgen)"
  pushd "$(mktemp -d --suffix dependencies)"
  npm i ngrok
  npx ngrok authtoken "$NGROK_TOKEN"
  (node <<TUNNEL
  const fs = require('fs');
  const ngrok = require('ngrok');
  (async function() {
    const jupyter = await ngrok.connect(8888);
    const tensorboard = await ngrok.connect(6006);
    const br = '\n*=*=*=*=*=*=*=*=*=*=*=*=*\n';
    fs.writeFileSync("log.md", \`\$${br}URL: Jupyter Lab: \$${jupyter}/lab?token=$${JUPYTER_TOKEN}\$${br}URL: Jupyter Notebook: \$${jupyter}/tree?token=$${JUPYTER_TOKEN}\$${br}URL: TensorBoard: \$${tensorboard}\$${br}\`);
  })();
  TUNNEL
  ) &
  while test ! -f log.md; do sleep 1; done
  cat log.md
  popd # dependencies

  # start tensorboard in background
  env -u JUPYTER_TOKEN -u AWS_ACCESS_KEY_ID -u AWS_SECRET_ACCESS_KEY -u REPO_TOKEN tensorboard --logdir . --host 0.0.0.0 --port 6006 &

  # start Jupyter server in foreground
  env -u AWS_ACCESS_KEY_ID -u AWS_SECRET_ACCESS_KEY -u REPO_TOKEN jupyter lab --allow-root --ip=0.0.0.0 --no-browser --port=8888 --port-retries=0
END
```

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+}
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:260]: no need to close a long-forgotten previous snippet?

```suggestion
```

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+```
+
+The second-to-last part of `main.tf` contains the script that will run once the
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:263]:
```suggestion
This part of our `main.tf` is the script to run once the
```

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+instance has been provisioned. These are CLI commands as you would run them in
+your local terminal. We won't discuss every individual line, but in broad
+strokes, the script does the following:
+
+1. Run a few configuration commands
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:268]: would drop this

```suggestion
```

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+1. Install the dependencies we will be needing
+1. Start an ngrok tunnel so that we can easily access the server through our
+   browser
+1. Launch the Jupyter server
+
+<admon type="info">
+
+This script also launches
+[Tensorboard](https://www.tensorflow.org/tensorboard/get_started) to run
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:277]:
```suggestion
[TensorBoard](https://www.tensorflow.org/tensorboard/get_started) to run
```

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+alongside Jupyter. It's not the main focus of this blog post, but it might be
+helpful!
+
+If you don't want Tensorboard running, simply remove or comment out
+[line 78 of `main.tf`](https://github.com/iterative/blog-tpi-jupyter/blob/aws/main.tf#L78).
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:282]:
```suggestion
If you don't want TensorBoard running or want to run some other web service instead,
simply remove or modify [the relevant line in `main.tf`](https://github.com/iterative/blog-tpi-jupyter/blob/e5fcc8aff74b40e1398ec0904efd73a2c480ff88/main.tf#L78).
```

exact lineno may change in future updates. Best to pin.

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+
+```bash
+output "urls" {
+  value = flatten(regexall("URL: (.+)", try(join("\n", iterative_task.jupyter_server.logs), "")))
+}
+```
+
+Lastly, we specify the outputs for our script. This is the value Terraform
+returns to be shown in our terminal after provisioning our instance. In this
+case, we are returning the URLs through which we can access the Jupyter server.
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:299]:
```suggestion
Lastly, purely for convenience, we specify the outputs of our task. Instead
of reading through all the script's output on the terminal, we extract the
ngrok tunnel URLs and inject them into an output value. Terraform will print
these outputs whenever we `refresh`. In this case, we get URLs to access the
Jupyter server.
```

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+
+# Next steps: rethink this approach
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:301]: this misses the point of the blog I think. It's not about "is Jupyter a good idea" it's more about "bring-your-own-cloud is cheaper and more configurable than depending on a comprehensive package solution."

See overall comments.

**Review by casperdcl — CHANGES_REQUESTED** [review]
Just re-iterating an earlier comment:

the point is to use Jupyter *as an example* to demonstrate the potential of TPI, it's not really about Jupyter itself at all. The main idea of TPI is to give you massive configurability without massive prerequisite knowledge. In general:

- **No CI/CD required**: You don’t need CI/CD to work in the cloud, you can get started right away
- **Simple one file setup**: install terraform, craft your `main.tf` and you are gewd to go
- **Extend your laptop hardware with the cloud**: "plug your cloud into your laptop" for easy access to those endless resources
- **Portable**: Share your `main.tf` and anyone can run the experiment. Also you if you have data/storage in the cloud already, you can use an instance/compute with the same cloud provider in the same cloud region to make access/egress faster & more secure (important for bug data & enterprise)
- **builtin spot recovery**: We handle instance rebooting and syncing data so you can easily take full advantage of spot instances

The Jupyter example showcases this by comparison to Google Colab. Were you there at the last office hours? CC @jendefig for the YouTube recording. The comparison is:

- :( more involved setup (download terraform, download our `main.tf`, sign up with AWS) compared to Colab's "just go to this website"
- :) configurable uptime
- :) configurable resources (CPU, GPU, RAM, storage)
- :) configurable pricing (bulk pricing of existing cloud provider, pay per-second/on-demand, make use of ~70% discount spot with TPI auto-spot recovery, etc.)

image

**Comment by casperdcl** [comment]
Merged some of the easier comments in a75886e, feel free to unresolve/revert/contest if you like.

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+# Under the hood
+
+"Alright, easy enough, but what is actually happening?"
+
+The major point of interest in the repository is
+[`main.tf`](https://github.com/iterative/blog-tpi-jupyter/blob/aws/main.tf).
+This file contains the specifications Terraform needs to provision our instance.
+Let's take a look at what each part of this configuration does:
+
+```
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:157]:
````suggestion
```hcl
````

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+terraform {
+  required_providers { iterative = { source = "iterative/iterative" } }
+}
+provider "iterative" {}
+```
+
+This first part of the configuration simply tells Terraform that we will be
+using TPI. We can consider it analogous to a package import in Python.
+
+```
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:167]:
````suggestion
```hcl
````

**Review by casperdcl — CHANGES_REQUESTED** [review]

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+```hcl
+resource "iterative_task" "jupyter_server" {
+  spot      = 0             # auto-priced low-cost spot instance
+  timeout   = 24*60*60      # force shutdown after 24h
+  disk_size = 125           # GB
+  machine   = "m+t4"        # m/l/xl (CPU), +k80/t4/v100 (GPU)
+  image     = "nvidia"      # or "ubuntu"
+
+  # cloud-specific config
+  cloud     = "aws"         # see `git checkout generic` branch for: gcp, az, k8s
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:176]:
```suggestion
  spot      = 0        # auto-priced low-cost spot instance
  timeout   = 24*60*60 # force shutdown after 24h
  disk_size = 125      # GB
  machine   = "m+t4"   # m/l/xl (CPU), +k80/t4/v100 (GPU)
  image     = "nvidia" # or "ubuntu"

  # cloud-specific config
  cloud     = "aws"    # see `git checkout generic` branch for: gcp, az, k8s
```

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+[the relevant line in `main.tf`](https://github.com/iterative/blog-tpi-jupyter/blob/e5fcc8aff74b40e1398ec0904efd73a2c480ff88/main.tf#L78).
+
+</admon>
+
+As you can see, the possibilities for the scripts we can run through TPI are
+extensive. We could add commands to clone a Git repository, for example. Or we
+could pull data in from a [DVC](https://dvc.org/) remote. This flexibility
+allows us to tailor the instance TPI provisions precisely to our needs.
+
+```bash
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:292]:
````suggestion
```hcl
````

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+  while test ! -f log.md; do sleep 1; done
+  cat log.md
+  popd # dependencies
+
+  # start tensorboard in background
+  env -u JUPYTER_TOKEN -u AWS_ACCESS_KEY_ID -u AWS_SECRET_ACCESS_KEY -u REPO_TOKEN tensorboard --logdir . --host 0.0.0.0 --port 6006 &
+
+  # start Jupyter server in foreground
+  env -u AWS_ACCESS_KEY_ID -u AWS_SECRET_ACCESS_KEY -u REPO_TOKEN jupyter lab --allow-root --ip=0.0.0.0 --no-browser --port=8888 --port-retries=0
+END
```
Comment (casperdcl) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:260]: for better syntax highlighting?

```suggestion
#!/bin/bash
set -euo pipefail
if test "$GITHUB_USER" != username; then
  # SSH debugging
  trap 'echo script error: waiting for debugging over SSH. Run \"terraform destroy\" to stop waiting; sleep inf' ERR
  mkdir -p "$HOME/.ssh"
  curl -fsSL "https://github.com/$GITHUB_USER.keys" >> "$HOME/.ssh/authorized_keys"
fi
export CUDACXX=/usr/local/cuda/bin/nvcc
export DEBIAN_FRONTEND=noninteractive
sed -ri 's#^(APT::Periodic::Unattended-Upgrade).*#\1 "0";#' /etc/apt/apt.conf.d/20auto-upgrades
dpkg-reconfigure unattended-upgrades
# install dependencies
pip3 install $${QUIET:+-q} jupyterlab notebook matplotlib ipywidgets tensorflow==2.8.0 tensorboard tensorflow_datasets
(curl -fsSL https://deb.nodesource.com/setup_16.x | bash -) >/dev/null
apt-get install -y $${QUIET:+-qq} nodejs

# start tunnel
export JUPYTER_TOKEN="$(uuidgen)"
pushd "$(mktemp -d --suffix dependencies)"
npm i ngrok
npx ngrok authtoken "$NGROK_TOKEN"
(node <<TUNNEL
const fs = require('fs');
const ngrok = require('ngrok');
(async function() {
  const jupyter = await ngrok.connect(8888);
  const tensorboard = await ngrok.connect(6006);
  const br = '\n*=*=*=*=*=*=*=*=*=*=*=*=*\n';
  fs.writeFileSync("log.md", \`\$${br}URL: Jupyter Lab: \$${jupyter}/lab?token=$${JUPYTER_TOKEN}\$${br}URL: Jupyter Notebook: \$${jupyter}/tree?token=$${JUPYTER_TOKEN}\$${br}URL: TensorBoard: \$${tensorboard}\$${br}\`);
})();
TUNNEL
) &
while test ! -f log.md; do sleep 1; done
cat log.md
popd # dependencies

# start tensorboard in background
env -u JUPYTER_TOKEN -u AWS_ACCESS_KEY_ID -u AWS_SECRET_ACCESS_KEY -u REPO_TOKEN tensorboard --logdir . --host 0.0.0.0 --port 6006 &

# start Jupyter server in foreground
env -u AWS_ACCESS_KEY_ID -u AWS_SECRET_ACCESS_KEY -u REPO_TOKEN jupyter lab --allow-root --ip=0.0.0.0 --no-browser --port=8888 --port-retries=0
```

**Review by casperdcl — CHANGES_REQUESTED** [review]

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 16]
```diff
@@ -0,0 +1,380 @@
...
+apt-get install -y $${QUIET:+-qq} nodejs
+
+# start tunnel
+export JUPYTER_TOKEN="$(uuidgen)"
+pushd "$(mktemp -d --suffix dependencies)"
+npm i ngrok
+npx ngrok authtoken "$NGROK_TOKEN"
+(node <<TUNNEL
+const fs = require('fs');
+const ngrok = require('ngrok');
```
Comment (jendefig) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:241]: This whole code section is where I get pretty much completely lost. While there are some comments put in which is helpful, there are other parts that are not explained.  What is CUDACXX? DEBIAN_FRONTEND?

Will this all look happy and understandable to a data scientist used to their Jupyter Notebook?  Or do we need to do some more work here?
Is this script interchangeable for all use cases?  What parts will need to be changed depending on their needs?

**Review by jendefig — CHANGES_REQUESTED** [review]
Looking good!  Some work to do.  But I like the way it's written.  I like the "flavor" you add to it.

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 17]
```diff
@@ -0,0 +1,380 @@
...
+charges to your credit card.
+
+</admon>
+
+# Guide
+
+Once we have cloned the repository and navigated to it in our terminal, we only
+need to run two commands. First, we use `terraform init` to initialize our
+configuration files. After that, we simply run `terraform apply` to launch our
+Jupyter server. That's all there as to it. TPI works its magic, and within a few
```
[the remark this answers is not in the document]
Comment (casperdcl) [reply content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:115]:
```suggestion
Jupyter server. That's all there is to it. TPI works its magic, and within a few
```

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md (hunk 16 above)
Comment (casperdcl) [reply content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:241]: I feel we don't need to dump the whole script here. Could condense it a lot. For a start:

```suggestion
#!/bin/bash
# install dependencies
pip install jupyterlab notebook matplotlib ipywidgets tensorflow tensorboard tensorflow_datasets
apt-get install -yq nodejs
npm i ngrok

# start tunnel
export JUPYTER_TOKEN="$(uuidgen)"
npx ngrok authtoken "$NGROK_TOKEN"
(node <<TUNNEL
const fs = require('fs');
const ngrok = require('ngrok');
```

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+environment, type `jupyter notebook` in our terminal, and head over to
+http://localhost:8888. We're good to go and ready to start prototyping!
+
+This usage has one major downside, however: we're limited by the specs of our
+local machine. While my laptop won't struggle too much training a classifier for
+the IRIS dataset, it might become a limiting factor when working on an
+alternative to GPT-3. And probably a little while before a project of that
+scope.
+
+The apparent solution to bypassing the limitations of our local machine is to
```
[the remark this answers is not in the document]
Comment (RCdeWit) [reply content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:44]: [Looked it up](https://www.google.com/search?q=apparent), I think I'm using the first definition of the word and you the second:

```
1. clearly visible or understood; obvious.
2. seeming real or true, but not necessarily so.
```

Will tweak slightly.

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 19]
```diff
@@ -0,0 +1,359 @@
...
+date: 2022-06-01
+description: >
+  Use TPI to quickly and cheaply provision a cloud-hosted Jupyter server to run
+  your notebooks on.
+descriptionLong: >
+  We can use TPI to quickly and cheaply provision cloud instances. This guide
+  will explore how to use TPI to launch a Jupyter server on an Amazon Web
+  Services EC2 instance at the lowest possible cost. With minor modifications,
+  you can also launch instances on Google Cloud Platform, Microsoft Azure, or
+  Kubernetes.
```
Comment (jendefig) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:12]: For the description, I think it would be good to include @casperdcl 's "bring your own cloud" catchphrase.  So after the first sentence, maybe:
This bring-your-own-cloud approach can launch instances on Google Cloud, Microsoft Azure, or Kubernetes with minor modifications.

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 20]
```diff
@@ -0,0 +1,379 @@
...
+problematic."
+
+Here are three of the limitations imposed by SaaS solutions such as Colab and
+SageMaker:
+
+- **Cost:** they are expensive to use. You're not only paying for the hardware
+  but also for using the platform.
+- **Walled garden:** they tend to be baked into the vendor's ecosystem. It's
+  difficult and sometimes impossible to use infrastructure components from
+  different vendors in conjunction with each other.
```
[the remark this answers is not in the document]
Comment (jendefig) [reply content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:67]: Have we done any other work to determine what other terminology around this concept is used in the industry.  And re: @casperdcl's Infrastructure components comment, it would be good to insert a "such as ____ and _____" to clarify this.

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 21]
```diff
@@ -0,0 +1,359 @@
...
+Launching Jupyter workspaces is one such an application.
+
+Here are four benefits to TPI that might pique your interest:
+
+- **Free and open:** TPI is entirely open source and free to use.
+- **Lower cloud costs:** TPI automatically cleans up unused instances. It also
+  allows you to use spot instances, reducing your per-hour costs.
+- **Modular:** TPI can be used with a variety of cloud vendors, meaning you can
+  use it regardless of what the rest of your stack looks like.
+- **Developer-first experience:** {{is this really a benefit for target audience
```
Comment (jendefig) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:86]: Ya. Good point.  This is a go-to for the ML engineer set, but maybe not as much for this audience.  

It could be replaced by some/all of the benefits Casper outlined in his comments before:

>No CI/CD required: You don’t need CI/CD to work in the cloud, you can get started right away
>Simple one file setup: install terraform, craft your main.tf >and you are gewd to go
>Extend your laptop hardware with the cloud: "plug your cloud into your laptop" for easy access to those endless resources
>Portable: Share your main.tf and anyone can run the experiment. Also you if you have data/storage in the cloud already, you can use an instance/compute with the same cloud provider in the same cloud region to make access/egress faster & more secure (important for bug data & enterprise)
>builtin spot recovery: We handle instance rebooting and syncing data so you can easily take full advantage of spot instances

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 22]
```diff
@@ -0,0 +1,379 @@
...
+example, you can use Azure or GCP with minor modifications.
+
+# Prerequisites
+
+Before you get started, make sure to take care of the following:
+
+1. [Download the Terraform CLI tool](https://www.terraform.io/downloads)
+1. [Set up an AWS Account](https://aws.amazon.com/)
+1. [Install AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html)
+1. [Configure your AWS authentication credentials with `aws configure`](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html)
```
[the remark this answers is not in the document]
Comment (jendefig) [reply content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:100]: @RCdeWit 👆 This wasn't corrected.  @casperdcl It's just the 4th one that's not required, correct?

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 23]
```diff
@@ -0,0 +1,379 @@
...
+1. [Get a free ngrok account for port forwarding](https://ngrok.com/)
+1. [Set your ngrok access token as an environment variable with `export NGROK_TOKEN="..."`](https://www.twilio.com/blog/2017/01/how-to-set-environment-variables.html)
+1. [Clone the example repository](https://github.com/iterative/blog-tpi-jupyter)
+
+<admon type="warn">
```
[the remark this answers is not in the document]
Comment (jendefig) [reply content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:105]: @RCdeWit 👆 Include in the admonition

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 24]
```diff
@@ -0,0 +1,379 @@
...
+[understand AWS pricing](https://aws.amazon.com/ec2/pricing/) to avoid unwelcome
+charges to your credit card.
+
+</admon>
+
+# Guide
+
+Once we have cloned the repository and navigated to it in our terminal, we only
+need to run two commands. First, we use `terraform init` to initialize our
+configuration files. After that, we simply run `terraform apply` to launch our
```
[the remark this answers is not in the document]
Comment (jendefig) [reply content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:119]: @RCdeWit 👆 to include

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 25]
```diff
@@ -0,0 +1,379 @@
...
+  environment = { NGROK_TOKEN = "", TF_CPP_MIN_LOG_LEVEL = "1", QUIET = "1", GITHUB_USER = "username" }
+  storage {
+    workdir = "shared"
+    output  = "."
+  }
+```
+
+The second part of the configuration is perhaps the most interesting. Here we
+provide the specifications for the instance Terraform should provision for us.
+In this case, we would like a medium CPU and an NVIDIA T4 GPU, along with a
```
[the remark this answers is not in the document]
Comment (jendefig) [reply content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:193]: You go over the sizes in the next part, but the link to the NIVIDIA t4 specs would be a good idea.

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md (hunk 16 above)
Comment (jendefig) [reply content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:250]: So I guess this needs to be corrected too. ☝️

**Review by jendefig — CHANGES_REQUESTED** [review]
Definitely some more to do.  Not all of Casper's comments/script adjustments were included.  But other wise, it's looking good.  The changes to the end are MUCH better. 👍 Almost there!

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (RCdeWit) [reply content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:67]: Will insert a phrase with examples.

With regards to `walled garden`: I see the term come up quite often (didn't invent it myself, after all). It's difficult to assess how ubiquitous this is; we can't Google how many people don't know something.

I think that if people don't immediately understand it, they can derive its meaning from context. But we can change the phrase if you think it makes the message clearer.

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (RCdeWit) [reply content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:101]: Correct, still working on this iteration.

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md [hunk 26]
```diff
@@ -0,0 +1,329 @@
...
+</admon>
+
+After specifying the details for the hardware, we configure a few environment
+variables for the instance. We also set the local working directory for the
+script and the directory to download files from (relative to the working
+directory) upon terminating the instance. In this case, both are the `shared`
+directory.
+
+```bash
+#!/bin/bash
```
Comment (RCdeWit) [inline content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:232]: @casperdcl Should we also update the example repo to reflect this change?

File: content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md (hunk 21 above)
Comment (RCdeWit) [reply content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md:86]: Did another iteration on this with the target audience in mind. The upsides now mirror the downsides of Colab etc. as listed above. 😄

**Review by casperdcl — CHANGES_REQUESTED** [review]
srry just took a few steps back & reviewed [the TPI release plan](https://github.com/orgs/iterative/projects/310); now questioning my sanity :sweat_smile: ... what is the actual reason for this blog post? It wasn't in the plan afaik. The complete blog plan was:

- Release announcement https://dvc.org/blog/terraform-provider
- Basic tutorial https://dvc.org/blog/local-experiments-to-cloud-with-tpi
- Docker tutorial https://dvc.org/blog/local-experiments-to-cloud-with-tpi-docker
- Jupyter & TensorBoard tutorial https://hackernoon.com/using-jupytertensorboard-in-any-cloud-with-one-command (full) & https://dev.to/casperdcl/how-to-run-jupyter-in-any-cloud-with-one-command-52me (short)

So... what's this post really about? Is it meant to say anything new? If not it may be best to [brutally cut our losses](https://hackernoon.com/combatting-overengineering-dont-be-afraid-to-throw-away-your-code-5e4235iq) (close unmerged).

**Comment by casperdcl** [comment]
Ah looks like some miscommunication, sorry didn't realise "it was supposed to be a quick retelling."

1) the entire goal of the original tutorial was to be posted on an external site *and not dvc.org*
2) if the original (external) posts are missing text/context, we can update them
3) it is exceedingly bad for SEO to re-post a slightly tweaked existing blog, esp. since we don't currently do [canonical links](https://en.wikipedia.org/wiki/Canonical_link_element)
4) it's also not worth our time to post non-original content

To be clear:

> quick retelling of the Hackernoon post [...] don't see why we should scrap the post unless you have major objections. It's as good as done, and even if doesn't add much, it surely doesn't hurt either.

- quick retelling of Hackernoon: already done in Dev.to
- I (and more importantly search engines) will not categorise this as "scrapping" - it's more like "not duplicating/plagiarising" (see point 3 above)
- I agree the original (Hackernoon) post could be improved (despite being thoroughly reviewed and iterated on by half a dozen people in detail)

TL;DR yes major objection, it will probably hurt.

I still think it's been a useful process, we can port over useful suggestions/improvements to the original post(s)?

**Comment by RCdeWit** [comment]
> Ah looks like some miscommunication, sorry didn't realise "it was supposed to be a quick retelling."
> 
> 1. the entire goal of the original tutorial was to be posted on an external site _and not dvc.org_
> 2. if the original (external) posts are missing text/context, we can update them
> 3. it is exceedingly bad for SEO to re-post a slightly tweaked existing blog, esp. since we don't currently do [canonical links](https://en.wikipedia.org/wiki/Canonical_link_element)
> 4. it's also not worth our time to post non-original content

Point-wise reply for clarity:

1. Yes, and the entire goal of this derivative is to be posted on dvc.org rather than an external site. Different outlets can serve different purposes. I can imagine Hackernoon was chosen in part for their platform and the people who visit that site. I can also imagine people searching our own outlets (Blog, docs, Youtube, ...) if they're looking for something specific. A post on dvc.org provides an extra opportunity for people to discover they can do this. Moreover, I think the target audience varies; here I tried to cater to more junior data scientists than the ones who might read the original post.
2. I think the original blog post is targeted well and serves its purpose; I don't see the immediate need to update it.
3. It is _not_ a slightly tweaked blog when looking at the text; it's a near-total rewrite. Conceptually it covers the same ground, but it does so with new content. It uses the original example you created, but the elaboration around is completely new. I don't see how this would have such a severe SEO impact.
4. To an extent I agree, although as outlined in [1] I believe these two posts should serve different purposes. Regardless, we're at the point where we spent the time anyway. We should definitely pay more mind to this in future instances, but at this point, we just need to dot the i's and cross the t's.

Just to clarify: with quick retelling I meant quick for the process, not the post itself. This post is far more verbose than your original publication.

**Comment by shcheklein** [comment]
@RCdeWit hey, any updates on this blog post? do we plan to publish or close this?
